src/domain/core/src/services/dependency_graph_service.rs

```diff
@@ -7,75 +7,111 @@
 // the Business Source License, use of this software will be governed
 // by the Apache License, Version 2.0.
 
-use internal_error::InternalError;
+use std::pin::Pin;
+use std::sync::Arc;
+
+use internal_error::{InternalError, ResultIntoInternal};
 use opendatafabric::DatasetID;
 use thiserror::Error;
 use tokio_stream::Stream;
 
+use crate::{DatasetRepository, GetSummaryOpts};
+
 /////////////////////////////////////////////////////////////////////////////////////////
 
 #[async_trait::async_trait]
 pub trait DependencyGraphService: Sync + Send {
+    /// Forces initialization of graph data, if it wasn't initialized already.
+    /// Ignored if called multiple times
+    async fn eager_initialization(
+        &self,
+        initializer: &DependencyGraphServiceInitializer,
+    ) -> Result<(), InternalError>;
+
     /// Iterates over 1st level of dataset's downstream dependencies
     async fn get_downstream_dependencies(
         &self,
         dataset_id: &DatasetID,
     ) -> Result<DatasetIDStream, GetDownstreamDependenciesError>;
 
-    /// Tracks a dependency between upstream and downstream dataset
-    ///
-    /// TODO: connect to event bus
-    async fn add_dependency(
+    /// Iterates over 1st level of dataset's upstream dependencies
+    async fn get_upstream_dependencies(
         &self,
-        dataset_upstream_id: &DatasetID,
-        dataset_downstream_id: &DatasetID,
-    ) -> Result<(), AddDependencyError>;
-
-    /// Removes tracked dependency between updstream and downstream dataset
-    ///
-    /// TODO: connect to event bus
-    async fn remove_dependency(
-        &self,
-        dataset_upstream_id: &DatasetID,
-        dataset_downstream_id: &DatasetID,
-    ) -> Result<(), RemoveDependencyError>;
+        dataset_id: &DatasetID,
+    ) -> Result<DatasetIDStream, GetUpstreamDependenciesError>;
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+#[dill::component(pub)]
+pub struct DependencyGraphServiceInitializer {
+    dataset_repo: Arc<dyn DatasetRepository>,
+}
+
+impl DependencyGraphServiceInitializer {
+    pub fn new(dataset_repo: Arc<dyn DatasetRepository>) -> Self {
+        Self { dataset_repo }
+    }
+
+    pub fn browse_dependencies_of_all_datasets(&self) -> DatasetDependenciesIDStream {
```

Comment:
`list_` would be better than `browse_`

again, trying to limit our function naming vocabulary to a base minimum

Reply:
Renamed